### DISCUSSION TITLE
Optional Path Parameters

### DISCUSSION THREAD

**Discussion by FallingSnow** [salvo-rs/salvo#397] [Q&A] [answered]
I couldn't find any examples/documentation on optional path parameters.

I have an endpoint that accepts an optional `filename` parameter but I don't how to tell salvo it is optional.

```rust
Router::with_path("api").push(Router::with_path("files/<filename>").handle(storage))
```

I've tried the follow:
```rust
Router::with_path("api").push(Router::with_path("files/<*filename>").handle(storage))
```
```rust
Router::with_path("api").push(Router::with_path("files/<?filename>").handle(storage))
```

But none of these paths are used when visiting `/api/files`. How do you specify a path parameter is optional?

**Comment by chrislearn** [comment, marked as answer]
Try <**filename>

**Reply by FallingSnow** [reply]
That works. Thank you!

I had to make a slight alteration to my get parameter code by adding a length filter:
```rust
req
    .params()
    .get("**filename")
    .filter(|s| s.len() > 0)
    .ok_or::<salvo::Error>(StatusError::bad_request().brief("missing filename").into())?
    .to_string()
```
